mm2src/coins/sia.rs

```diff
@@ -0,0 +1,564 @@
+            TradeFee, TransactionEnum, TransactionFut};
+use crate::{coin_errors::MyAddressError, BalanceFut, CanRefundHtlc, CheckIfMyPaymentSentArgs, CoinFutSpawner,
+            ConfirmPaymentInput, FeeApproxStage, FoundSwapTxSpend, MakerSwapTakerCoin, MmCoinEnum,
+            NegotiateSwapContractAddrErr, PaymentInstructionArgs, PaymentInstructions, PaymentInstructionsErr,
+            PrivKeyBuildPolicy, RawTransactionResult, RefundPaymentArgs, RefundResult, SearchForSwapTxSpendInput,
+            SendMakerPaymentSpendPreimageInput, SendPaymentArgs, SignRawTransactionRequest, SignatureResult,
+            SpendPaymentArgs, TakerSwapMakerCoin, TradePreimageFut, TradePreimageResult, TradePreimageValue,
+            TransactionResult, TxMarshalingErr, UnexpectedDerivationMethod, ValidateAddressResult, ValidateFeeArgs,
+            ValidateInstructionsErr, ValidateOtherPubKeyErr, ValidatePaymentError, ValidatePaymentFut,
+            ValidatePaymentInput, ValidatePaymentResult, VerificationResult, WaitForHTLCTxSpendArgs, WatcherOps,
+            WatcherReward, WatcherRewardError, WatcherSearchForSwapTxSpendInput, WatcherValidatePaymentInput,
+            WatcherValidateTakerFeeInput, WithdrawFut, WithdrawRequest};
+use crate::{DexFee, ValidateWatcherSpendInput};
+use async_trait::async_trait;
+use common::executor::AbortedError;
+use futures::{FutureExt, TryFutureExt};
+use futures01::Future;
+use keys::KeyPair;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::*;
+use mm2_number::{BigDecimal, MmNumber};
+use rpc::v1::types::Bytes as BytesJson;
+use serde_json::Value as Json;
+use std::ops::Deref;
+use std::sync::Arc;
+
+pub mod http_client;
+use http_client::{SiaApiClient, SiaApiClientError};
+
+use url::Url;
+
+use mm2_core::mm_ctx::MmWeak;
+
+#[derive(Clone)]
+pub struct SiaCoin(SiaArc);
+#[derive(Clone)]
+pub struct SiaArc(Arc<SiaCoinFields>);
+
+#[derive(Debug, Display)]
+pub enum SiaConfError {
+    #[display(fmt = "'foo' field is not found in config")]
+    Foo,
+    Bar(String),
+}
+
+pub type SiaConfResult<T> = Result<T, MmError<SiaConfError>>;
+
+#[derive(Debug)]
+pub struct SiaCoinConf {
+    ticker: String,
+    pub foo: u32,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct SiaCoinActivationParams {
+    #[serde(default)]
+    pub tx_history: bool,
+    pub required_confirmations: Option<u64>,
+    pub gap_limit: Option<u32>,
+    pub http_url: Url,
+    pub http_auth: String,
+}
+
+pub struct SiaConfBuilder<'a> {
+    #[allow(dead_code)]
+    conf: &'a Json,
+    ticker: &'a str,
+}
+
+impl<'a> SiaConfBuilder<'a> {
+    pub fn new(conf: &'a Json, ticker: &'a str) -> Self { SiaConfBuilder { conf, ticker } }
+
+    pub fn build(&self) -> SiaConfResult<SiaCoinConf> {
+        Ok(SiaCoinConf {
+            ticker: self.ticker.to_owned(),
+            foo: 0,
+        })
+    }
+}
+
+pub struct SiaCoinFields {
```

Comment:
Don't forget to add other fields that might be needed, I don't know if sia will need the following or not:
- tx_fee
- dust_amount
- derivation_method (this will be needed for sure once HD wallet PRs are merged)
- tx_cache
- recently_spent_outpoints
- check_utxo_maturity
- abortable_system (will be needed for sure)
- block_headers_status_notifier / block_headers_status_watcher (will be needed for SPV)

Reply:
Prefer to add them as needed to reduce unused code clutter. Adding a code comment referencing your feedback.